java/cuvs-java/src/main/java22/com/nvidia/cuvs/spi/NativeDependencyLoader.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.nvidia.cuvs.spi;
+
+import static com.nvidia.cuvs.internal.common.NativeLibraryUtils.JVM_LoadLibrary$mh;
+
+import java.io.*;
+import java.lang.foreign.Arena;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.StandardCopyOption;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+
+/**
+ * A class that loads native dependencies if they are available in the jar.
+ */
+class NativeDependencyLoader {
+
+  interface NativeDependencyLoaderStrategy {
+    void loadLibraries() throws ProviderInitializationException;
+  }
+
+  private static final NativeDependencyLoaderStrategy LOADER_STRATEGY = createLoaderStrategy();
+
+  private static NativeDependencyLoaderStrategy createLoaderStrategy() {
+    if (jarHasNativeDependencies()) {
+      return new EmbeddedNativeDependencyLoaderStrategy();
+    } else {
+      return new SystemNativeDependencyLoaderStrategy();
+    }
+  }
+
+  private static boolean jarHasNativeDependencies() {
+    try (var jarFile =
+        new JarFile(
+            JDKProvider.class.getProtectionDomain().getCodeSource().getLocation().getPath())) {
+      Manifest manifest = jarFile.getManifest();
+      // TODO: use this variable to add a check on the installed CUDA version
+      // (which will be system-loaded in any case, even with the fat-jar)
+      var embeddedLibrariesCudaVersion =
+          manifest.getMainAttributes().getValue("Embedded-Libraries-Cuda-Version");
+      return embeddedLibrariesCudaVersion != null;
+    } catch (IOException e) {
+      return false;
+    }
+  }
+
+  private static boolean loaded = false;
+
+  static void loadLibraries() throws ProviderInitializationException {
+    if (!loaded) {
+      try {
+        LOADER_STRATEGY.loadLibraries();
```

Comment:
@chatman I think pre-loading libraries this way will already solve the problem with cudart, but if it does not we can add the System.loadLibrary around here (e.g. in the implementation of `loadLibraries`), then also call `cudaRuntimeGetVersion` and compare it with `embeddedLibrariesCudaVersion` for an additional check. 
Or the other way around: use `embeddedLibrariesCudaVersion` to load a specific version of cudart

(All of this in a follow-up I think)